include/caffe/util/thread_pool.hpp

```diff
@@ -0,0 +1,112 @@
+#ifndef THREAD_POOL_H
```

Comment:
I actually tried to start that way since I don't enjoy reimplementing things.  I kept running into issues with it not working how I needed, explicitly the notion of completion of work.  Or another way to say it is that to use boost's thread_group I had to almost write the same amount of code.